### ISSUE TITLE
Define fixed values for xlink:{actuate,type}.

### ISSUE THREAD

**Issue by skynavga** [w3c/ttml2#1039]
Even though TTML2 does not specify the use of `xlink:actuate` or `xlink:type`, their semantics (as defined by [XLink]) still apply, and, in particular, their values (and consequent semantics) are presently defined in the TTML2 schemas as `#fixed`, with values:

```
xlink:actuate = "onRequest"
xlink:type    = "simple"
```

**Comment by nigelmegitt** [comment]
The spec for [`actuate`](https://www.w3.org/TR/xlink11/#link-behaviors) says that when absent the implied behaviour is `none`, so that's what I assumed applied here, i.e. TTML2 doesn't specify it. That doesn't seem to need a change, since we haven't got any specific requirements or reasons to constrain implementations to a particular actuation behaviour.

I agree that it's implied but not stated that `xlink:type` is `simple`.

**Comment by skynavga** [comment]
You misread the XLink spec. The DTD syntax

```
xlink:actuate   (onLoad
                |onRequest
                |other
                |none)          #IMPLIED
```

does not mean that `none` is implied. See [XML 1.1 §3.3.2](https://www.w3.org/TR/2006/REC-xml11-20060816/#sec-attr-defaults)

>In an attribute declaration, #REQUIRED means that the attribute must always be provided, #IMPLIED that no default value is provided.

Try reading the syntax fragment as

```
<!ATTLIST simplelink xlink:actuate (onLoad|onRequest|other|none) #IMPLIED>
```

So we DO need to specify a default for `xlink:show` as well as `xlink:type`.

**Comment by nigelmegitt** [comment]
Thanks @skynavga I did indeed misread it by not noticing the `)`.

I still think `none` would be a good choice though.

**Comment by skynavga** [comment]
Take note that SVG 1.1, which uses XLink 1.1, defines attributes on the `a` element as

```
<!ATTLIST %SVG.a.qname;
    %SVG.Core.attrib;
    %SVG.Conditional.attrib;
    %SVG.Style.attrib;
    %SVG.Presentation.attrib;
    %SVG.GraphicalEvents.attrib;
    %SVG.XLinkReplace.attrib;
    %SVG.External.attrib;
    transform %TransformList.datatype; #IMPLIED
    target %LinkTarget.datatype; #IMPLIED
>
```

where

<!ENTITY % SVG.XLinkReplace.attrib
    "%XLINK.xmlns.attrib;
     %XLINK.pfx;type ( simple ) #FIXED 'simple'
     %XLINK.pfx;href %URI.datatype; #REQUIRED
     %XLINK.pfx;role %URI.datatype; #IMPLIED
     %XLINK.pfx;arcrole %URI.datatype; #IMPLIED
     %XLINK.pfx;title CDATA #IMPLIED
     %XLINK.pfx;show ( new | replace ) 'replace'
     %XLINK.pfx;actuate ( onRequest ) #FIXED 'onRequest'
     %SVG.XLinkReplace.extra.attrib;"
>

so SVG uses `onRequest` as a fixed value for `xlink:actuate`.

We should do the same.

**Comment by nigelmegitt** [comment]
What's the overlap in requirement here between TTML and SVG? 

As I read it, if an application supports the `onRequest` behaviour that would be reasonable if TTML2 specifies `none`, so it would not be incompatible from that perspective.

Whereas if TTML2 specifies `onRequest` then applications are not permitted to do any other behaviour. Given that we don't have a particular use case in mind, I think it would be too strict.

An example possible use case that would be reasonable in the context of TTML2 and would not be compatible with `onRequest` is one where a media presentation triggers synchronised events by loading the linked URLs, say to trigger a second screen to provide information. This may need `other` behaviour. Or the act of playing a TTML2 file might open another application somewhere, which would need `onLoad` behaviour.

**Comment by skynavga** [comment]
The overlap with SVG is that we are defining the same semantics for link usage as employed with `svg:a`. We are not defining other possible link semantic scenarios such as you mention that might be associated with different element types or usage contexts in TTML. In SVG, `xlink:actuate` is fixed according to the element type:

- `svg:a` uses `onRequest`
- `svg:cursor`,  `svg:feImage`, `svg:image`, `svg:mpath`, `svg:textpath`, `svg:tref`, and `svg:use` use `onLoad`

At present, in TTML2, our use of `xlink:*` on `tt:image` and `tt:span` correspond with the `svg:a` semantics, and thus dictate the use of `onRequest`. In the future (TTML3, etc) we can define other semantics that may dictate the use of `onLoad` or `none`.

So my answer to your query is, yes, we do want all `xlink:*` actuations to be `onRequest` in TTML2.

**Comment by nigelmegitt** [comment]
Sorry for not coming back to this for such a long time - revisiting due to the pull request.

> The overlap with SVG is that we are defining the same semantics for link usage as employed with `svg:a`. We are not defining other possible link semantic scenarios such as you mention that might be associated with different element types or usage contexts in TTML.

I strongly disagree. In fact we don't define the semantics at all. On both an `image` and a `span` element it would be entirely legitimate right now to do any of:

* use as a "clickable" link that is actionable post-load (`onRequest`) by the user or the application under some setting, 
* to trigger on load (`onLoad`) and embed the referenced content in place of the element content (I'd accept that this should be done by defining the URL on an image resource but we can't do it for a text resource right now),
* to trigger on load (`onLoad`) and not modify the content locally at all but instead use the traversal request to trigger behaviour in another context, such as a remote device, another window on the same device etc.

In summary, defining it as `none` doesn't seem ideal, but defining it as `onRequest` prevents some use cases that could currently be permitted, and so would be a breaking change.

**Comment by skynavga** [comment]
Here's the problem: we defined the default for `xlink:show` as `new`, not `none`. We need to specify either:

(1) `xlink:show` defaults to `new` and `xlink:actuate` has an implied `onRequest`;

or

(2) `xlink:show` defaults to `none` and `xlink:actuate` has an implied `none`;

When I originally specified the `xlink:*` attributes, I definitely intended option (1) to apply, but failed to document the implied value for `xlink:actuate` as `onRequest`.

Now, you are suggesting we head in a different direction, one that requires making a backward incompatible change to `xlink:show` default. I would prefer we finish what I intended, and then, if we want, we can add `xlink:actuate` as a defined attribute in TTML3, which opens up all the possibilities that you suggest.

**Comment by skynavga** [comment]
I suppose another option (3) in TTML2 2e is to say that `xlink:actuate` has implied `onRequest` semantics if `xlink:show`'s computed value (specified or defaulted) is `new`, and, otherwise, has implied `none` semantics.

**Comment by nigelmegitt** [comment]
I've given this some more thought over the weekend and changed my mind about what we should do. 

Firstly, I do not know of an actual request for the `onLoad` functionality at this time, and if there is one, then we can address that later. 

Secondly, what I know of multi-screen synchronisation suggests that attempting to use TTML playback as the mechanism will be sub-optimal. Additionally, even if not used for synchronisation, the `onLoad` mode could be used for other purposes, such as tracking media playback. Again, this would be a sub-optimal solution; most players would integrate this directly rather than relying on triggers in the document being played. I also realised this would have implications on security and privacy that we (I!) have not yet thought through.

For example if we allow `onLoad` and `embed` this presents the possibility of:

```
<span xlink:show="embed" xlink:actuate="onLoad" xlink:href="[url_1]">Embed in place of this</span>
```
then at url1:
```
<span xlink:show="embed" xlink:actuate="onLoad" xlink:href="[url_2]">This is what url_1 gives</span>
```
then at url2:
```
<span xlink:show="embed" xlink:actuate="onLoad" xlink:href="[url_1]">url_2 sends you back to url_1 on load</span>
```

When does the `onLoad` loop end?


I like the idea of @skynavga 's suggestion at https://github.com/w3c/ttml2/issues/1039#issuecomment-475815774 as a way through this though. A further alternative given the possible issues (as above) with `xlink:show="embed"` or `xlink:show="replace"` is actually to remove those as possible values - we should seek evidence of usage before doing that though, since it would effectively be a breaking change.
